This exercise will require you to pull some data from https://data.nasdaq.com/ (formerly Quandl API).

As a first step, you will need to register a free account on the https://data.nasdaq.com/ website.

After you register, you will be provided with a unique API key, that you should store:

*Note*: Use a `.env` file and put your key in there and `python-dotenv` to access it in this notebook. 

The code below uses a key that was used when generating this project but has since been deleted. Never submit your keys to source control. There is a `.env-example` file in this repository to illusrtate what you need. Copy that to a file called `.env` and use your own api key in that `.env` file. Make sure you also have a `.gitignore` file with a line for `.env` added to it. 

The standard Python gitignore is [here](https://github.com/github/gitignore/blob/master/Python.gitignore) you can just copy that. 

In [1]:
# get api key from your .env file
import os
from dotenv import load_dotenv
from collections import defaultdict

load_dotenv()
API_KEY = os.getenv('NASDAQ_API_KEY')

print(API_KEY)

dddddddd


Nasdaq Data has a large number of data sources, but, unfortunately, most of them require a Premium subscription. Still, there are also a good number of free datasets.

For this mini project, we will focus on equities data from the Frankfurt Stock Exhange (FSE), which is available for free. We'll try and analyze the stock prices of a company called Carl Zeiss Meditec, which manufactures tools for eye examinations, as well as medical lasers for laser eye surgery: https://www.zeiss.com/meditec/int/home.html. The company is listed under the stock ticker AFX_X.

You can find the detailed Nasdaq Data API instructions here: https://docs.data.nasdaq.com/docs/in-depth-usage

While there is a dedicated Python package for connecting to the Nasdaq API, we would prefer that you use the *requests* package, which can be easily downloaded using *pip* or *conda*. You can find the documentation for the package here: http://docs.python-requests.org/en/master/ 

Finally, apart from the *requests* package, you are encouraged to not use any third party Python packages, such as *pandas*, and instead focus on what's available in the Python Standard Library (the *collections* module might come in handy: https://pymotw.com/3/collections/).
Also, since you won't have access to DataFrames, you are encouraged to us Python's native data structures - preferably dictionaries, though some questions can also be answered using lists.
You can read more on these data structures here: https://docs.python.org/3/tutorial/datastructures.html

Keep in mind that the JSON responses you will be getting from the API map almost one-to-one to Python's dictionaries. Unfortunately, they can be very nested, so make sure you read up on indexing dictionaries in the documentation provided above.

In [253]:
# First, import the relevant modules
import requests

Note: API's can change a bit with each version, for this exercise it is reccomended to use the nasdaq api at `https://data.nasdaq.com/api/v3/`. This is the same api as what used to be quandl so `https://www.quandl.com/api/v3/` should work too.

Hint: We are looking for the `AFX_X` data on the `datasets/FSE/` dataset.

In [254]:
# Now, call the Nasdaq API and pull out a small sample of the data (only one day) to get a glimpse
# into the JSON structure that will be returned
res = requests.get('https://data.nasdaq.com/api/v3/datasets/FSE/AFX_X/?&start_date=2014-11-01&end_date=2014-11-03?api_key=API_KEY')

In [255]:
# Inspect the JSON structure of the object you created, and take note of how nested it is,
# as well as the overall structure
json = res.json()
json['dataset']

{'id': 10095370,
 'dataset_code': 'AFX_X',
 'database_code': 'FSE',
 'name': 'Carl Zeiss Meditec (AFX_X)',
 'description': 'Stock Prices for Carl Zeiss Meditec (2020-11-02) from the Frankfurt Stock Exchange.<br><br>Trading System: Xetra<br><br>ISIN: DE0005313704',
 'refreshed_at': '2020-12-01T14:48:09.907Z',
 'newest_available_date': '2020-12-01',
 'oldest_available_date': '2000-06-07',
 'column_names': ['Date',
  'Open',
  'High',
  'Low',
  'Close',
  'Change',
  'Traded Volume',
  'Turnover',
  'Last Price of the Day',
  'Daily Traded Units',
  'Daily Turnover'],
 'frequency': 'daily',
 'type': 'Time Series',
 'premium': False,
 'limit': None,
 'transform': None,
 'column_index': None,
 'start_date': '2014-11-01',
 'end_date': '2014-11-03',
 'data': [['2014-11-03',
   21.86,
   21.86,
   21.43,
   21.67,
   None,
   17346.0,
   376461.0,
   None,
   None,
   None]],
 'collapse': None,
 'order': None,
 'database_id': 6129}

These are your tasks for this mini project:

1. Collect data from the Franfurt Stock Exchange, for the ticker AFX_X, for the whole year 2017 (keep in mind that the date format is YYYY-MM-DD).
2. Convert the returned JSON object into a Python dictionary.
3. Calculate what the highest and lowest opening prices were for the stock in this period.
4. What was the largest change in any one day (based on High and Low price)?
5. What was the largest change between any two days (based on Closing Price)?
6. What was the average daily trading volume during this year?
7. (Optional) What was the median trading volume during this year. (Note: you may need to implement your own function for calculating the median.)

In [256]:
res_2017 = requests.get('https://data.nasdaq.com/api/v3/datasets/FSE/AFX_X/?&start_date=2017-01-01&end_date=2017-12-30?api_key=API_KEY')
json_2017 = res_2017.json()

In [257]:
AFX_X_dict = defaultdict(str)

In [258]:
json_2017['dataset']['column_names'][0]

'Date'

In [259]:
json_2017['dataset']['data'][0][0]

'2017-12-29'

In [260]:
for i in range(len(json_2017['dataset']['column_names'])):
    AFX_X_dict[json_2017['dataset']['column_names'][i]] = [json_2017['dataset']['data'][0][i]]

In [261]:
AFX_X_dict

defaultdict(str,
            {'Date': ['2017-12-29'],
             'Open': [51.76],
             'High': [51.94],
             'Low': [51.45],
             'Close': [51.76],
             'Change': [None],
             'Traded Volume': [34640.0],
             'Turnover': [1792304.0],
             'Last Price of the Day': [None],
             'Daily Traded Units': [None],
             'Daily Turnover': [None]})

In [262]:
length = len(json_2017['dataset']['data'])

In [263]:
length

255

In [264]:
for i in range(len(json_2017['dataset']['column_names'])):
    for j in range(1,length):
        AFX_X_dict[json_2017['dataset']['column_names'][i]].append(json_2017['dataset']['data'][j][i])

In [265]:
AFX_X_dict['Open']

[51.76,
 51.65,
 51.45,
 51.05,
 51.16,
 51.88,
 52.73,
 52.37,
 52.7,
 53.11,
 52.64,
 52.29,
 52.28,
 51.5,
 50.89,
 50.8,
 51.21,
 49.5,
 49.52,
 48.64,
 49.64,
 49.09,
 49.13,
 49.11,
 48.8,
 48.4,
 47.25,
 46.57,
 47.03,
 47.09,
 47.98,
 48.4,
 48.38,
 47.3,
 47.65,
 46.42,
 46.16,
 45.81,
 45.0,
 45.88,
 46.29,
 46.53,
 45.48,
 45.2,
 45.01,
 45.16,
 44.9,
 45.08,
 45.72,
 46.01,
 45.8,
 45.61,
 45.5,
 45.58,
 45.97,
 45.64,
 46.2,
 46.19,
 46.01,
 45.36,
 44.51,
 43.58,
 42.0,
 42.35,
 42.3,
 42.3,
 41.48,
 42.29,
 42.54,
 42.65,
 42.5,
 42.29,
 42.35,
 42.49,
 43.21,
 42.81,
 42.7,
 43.0,
 42.66,
 43.0,
 42.38,
 42.16,
 42.0,
 42.0,
 41.71,
 42.11,
 42.64,
 42.72,
 42.82,
 42.46,
 42.42,
 42.28,
 41.88,
 42.4,
 42.53,
 42.12,
 41.3,
 41.73,
 43.5,
 44.9,
 45.85,
 45.13,
 45.34,
 45.25,
 45.24,
 44.94,
 45.26,
 45.16,
 44.91,
 44.7,
 45.31,
 45.57,
 45.74,
 45.06,
 45.5,
 45.6,
 45.07,
 44.67,
 44.29,
 44.94,
 44.64,
 44.79,
 45.5,
 44.67,
 45.83,
 45.29,
 45.01,
 45.73,
 46.68,

In [266]:
AFX_X_dict['Open'] = list(filter(None,AFX_X_dict['Open']))

In [267]:
AFX_X_dict['Open']

[51.76,
 51.65,
 51.45,
 51.05,
 51.16,
 51.88,
 52.73,
 52.37,
 52.7,
 53.11,
 52.64,
 52.29,
 52.28,
 51.5,
 50.89,
 50.8,
 51.21,
 49.5,
 49.52,
 48.64,
 49.64,
 49.09,
 49.13,
 49.11,
 48.8,
 48.4,
 47.25,
 46.57,
 47.03,
 47.09,
 47.98,
 48.4,
 48.38,
 47.3,
 47.65,
 46.42,
 46.16,
 45.81,
 45.0,
 45.88,
 46.29,
 46.53,
 45.48,
 45.2,
 45.01,
 45.16,
 44.9,
 45.08,
 45.72,
 46.01,
 45.8,
 45.61,
 45.5,
 45.58,
 45.97,
 45.64,
 46.2,
 46.19,
 46.01,
 45.36,
 44.51,
 43.58,
 42.0,
 42.35,
 42.3,
 42.3,
 41.48,
 42.29,
 42.54,
 42.65,
 42.5,
 42.29,
 42.35,
 42.49,
 43.21,
 42.81,
 42.7,
 43.0,
 42.66,
 43.0,
 42.38,
 42.16,
 42.0,
 42.0,
 41.71,
 42.11,
 42.64,
 42.72,
 42.82,
 42.46,
 42.42,
 42.28,
 41.88,
 42.4,
 42.53,
 42.12,
 41.3,
 41.73,
 43.5,
 44.9,
 45.85,
 45.13,
 45.34,
 45.25,
 45.24,
 44.94,
 45.26,
 45.16,
 44.91,
 44.7,
 45.31,
 45.57,
 45.74,
 45.06,
 45.5,
 45.6,
 45.07,
 44.67,
 44.29,
 44.94,
 44.64,
 44.79,
 45.5,
 44.67,
 45.83,
 45.29,
 45.01,
 45.73,
 46.68,

In [280]:
len(AFX_X_dict['Open'])

252

In [268]:
max_price = AFX_X_dict['Open'][0]
min_price = AFX_X_dict['Open'][0]
for i in range(len(AFX_X_dict['Open'])):
    if max_price < AFX_X_dict['Open'][i]:
        max_price = AFX_X_dict['Open'][i]
    if min_price > AFX_X_dict['Open'][i]:
        min_price = AFX_X_dict['Open'][i]
        

In [269]:
max_price, min_price

(53.11, 34.0)

So the highest open price is 53.11 and lowest open price is 34.0

In [270]:
largest_change_daily = AFX_X_dict["High"][0] - AFX_X_dict['Low'][0]
for i in range(length):
    if largest_change_daily < AFX_X_dict["High"][i] - AFX_X_dict['Low'][i]:
        largest_change_daily = AFX_X_dict["High"][i] - AFX_X_dict['Low'][i]



In [271]:
largest_change_daily

2.8100000000000023

So the largest daily change is 2.81

In [272]:
largest_change_btn_two_days = abs(AFX_X_dict['Close'][0] - AFX_X_dict['Close'][1])
for i in range(length-1):
    if largest_change_btn_two_days < abs(AFX_X_dict['Close'][i] - AFX_X_dict['Close'][i+1]):
        largest_change_btn_two_days = abs(AFX_X_dict['Close'][i] - AFX_X_dict['Close'][i+1])
    

In [273]:
largest_change_btn_two_days

2.559999999999995

So the largest change between any two days is 2.56

In [274]:
AFX_X_dict['Traded Volume']

[34640.0,
 40660.0,
 57452.0,
 71165.0,
 120649.0,
 50587.0,
 137313.0,
 129733.0,
 204080.0,
 132981.0,
 128434.0,
 87911.0,
 71817.0,
 109157.0,
 48123.0,
 88730.0,
 83023.0,
 94385.0,
 101733.0,
 123019.0,
 67342.0,
 42669.0,
 102180.0,
 50350.0,
 38834.0,
 91142.0,
 78502.0,
 97252.0,
 54107.0,
 89373.0,
 67593.0,
 67672.0,
 76286.0,
 90245.0,
 120268.0,
 94195.0,
 48152.0,
 60716.0,
 56911.0,
 37958.0,
 56319.0,
 56245.0,
 74472.0,
 56319.0,
 47730.0,
 43042.0,
 43375.0,
 55707.0,
 59991.0,
 45263.0,
 65837.0,
 49246.0,
 43362.0,
 49180.0,
 69455.0,
 65860.0,
 44059.0,
 66760.0,
 94804.0,
 115706.0,
 95313.0,
 99821.0,
 157234.0,
 76600.0,
 51321.0,
 56224.0,
 79955.0,
 105194.0,
 57838.0,
 65546.0,
 44037.0,
 107144.0,
 65157.0,
 68801.0,
 52828.0,
 103273.0,
 59881.0,
 64320.0,
 71006.0,
 66351.0,
 105288.0,
 151474.0,
 157888.0,
 97804.0,
 98156.0,
 47130.0,
 69734.0,
 65213.0,
 70269.0,
 95376.0,
 68812.0,
 72886.0,
 131361.0,
 104676.0,
 64334.0,
 127682.0,
 183412.0,
 175161

In [275]:
total_traded_volume = 0.0
for i in range(length):
    total_traded_volume = total_traded_volume + AFX_X_dict['Traded Volume'][i]
    
avg_traded_volume = total_traded_volume/length

In [276]:
avg_traded_volume

89124.33725490196

So the average traded volume is 89124.

In [277]:
for i in range(length):
    for j in range(i+1,length):
        if AFX_X_dict['Traded Volume'][i] > AFX_X_dict['Traded Volume'][j]:
            switch = AFX_X_dict['Traded Volume'][i]
            AFX_X_dict['Traded Volume'][i] = AFX_X_dict['Traded Volume'][j]
            AFX_X_dict['Traded Volume'][j] = switch
            

In [278]:
median_traded_volume = (AFX_X_dict['Traded Volume'][int(length/2)] + AFX_X_dict['Traded Volume'][int(length/2+1)]) / 2

In [279]:
median_traded_volume

76443.0

So median traded volume is 76443